display_list/display_list.cc

```diff
@@ -73,86 +89,114 @@ uint32_t DisplayList::next_unique_id() {
+        save_was_needed(save_was_needed) {}
+
+  DlIndex previous_restore_index;
+  bool save_was_needed;
 };
-class NopCuller final : public Culler {
- public:
-  static NopCuller instance;
-
-  ~NopCuller() = default;
-
-  bool init(DispatchContext& context) override {
-    // Setting next_render_index to 0 means that
-    // all rendering ops will be at or after that
-    // index so they will execute and all restore
-    // indices will be after it as well so all
-    // clip and transform operations will execute.
-    context.next_render_index = 0;
-    return true;
+
+void DisplayList::RTreeResultsToIndexVector(
+    std::vector<DlIndex>& indices,
+    const std::vector<int>& rtree_results) const {
+  FML_DCHECK(rtree_);
+  auto cur_rect = rtree_results.begin();
+  auto end_rect = rtree_results.end();
+  if (cur_rect >= end_rect) {
+    return;
   }
-  void update(DispatchContext& context) override {}
-};
-NopCuller NopCuller::instance = NopCuller();
-class VectorCuller final : public Culler {
- public:
-  VectorCuller(const DlRTree* rtree, const std::vector<int>& rect_indices)
-      : rtree_(rtree), cur_(rect_indices.begin()), end_(rect_indices.end()) {}
-
-  ~VectorCuller() = default;
-
-  bool init(DispatchContext& context) override {
-    if (cur_ < end_) {
-      context.next_render_index = rtree_->id(*cur_++);
-      return true;
-    } else {
-      // Setting next_render_index to MAX_INT means that
-      // all rendering ops will be "before" that index and
-      // they will skip themselves and all clip and transform
-      // ops will see that the next render index is not
-      // before the next restore index (even if both are MAX_INT)
-      // and so they will also not execute.
-      // None of this really matters because returning false
-      // here should cause the Dispatch operation to abort,
-      // but this value is conceptually correct if that short
-      // circuit optimization isn't used.
-      context.next_render_index = std::numeric_limits<int>::max();
-      return false;
+  DlIndex next_render_index = rtree_->id(*cur_rect++);
+  DlIndex next_restore_index = std::numeric_limits<DlIndex>::max();
+  std::vector<SaveInfo> save_infos;
+  for (DlIndex index = 0u; index < offsets_.size(); index++) {
+    while (index > next_render_index) {
+      if (cur_rect < end_rect) {
+        next_render_index = rtree_->id(*cur_rect++);
+      } else {
+        // Nothing left to render.
+        // Nothing left to do, but match our restores from the stack.
+        while (!save_infos.empty()) {
+          SaveInfo& info = save_infos.back();
+          // stack top boolean tells us whether the local variable
+          // next_restore_index should be executed. The local variable
+          // then gets reset to the value stored in the stack top
+          if (info.save_was_needed) {
+            FML_DCHECK(next_restore_index < offsets_.size());
+            indices.emplace_back(next_restore_index);
```

Comment:
Minor nit, but I would prefer push_back if we're not using the aggregate initialization features of emplace_back.

Reply:
There were other places that also used emplace_back with a simple integer type (all in this PR) so I changed them all. Only one emplace_back that was actually invoking an aggregate constructor.